src/Bicep.Core/Analyzers/Linter/Common/LinterExpressionHelper.cs

```diff
@@ -0,0 +1,148 @@
+        /// Attempts to find a resource with the same name as the given expression
+        /// </summary>
+        /// <remarks>This is not intended to be foolproof, but simply to handle common scenarios</remarks>
+        public static IEnumerable<DeclaredResourceMetadata> TryFindResourceByNameExpression(SemanticModel model, SyntaxBase resourceNameExpression)
+        {
+            // CONSIDER: Support child resources and arrays and other improvements
+
+            var resourcesAndNames = new List<(DeclaredResourceMetadata resource, SyntaxBase name)>();
+            foreach (var resource in model.DeclaredResources.Where(r => r.IsAzResource))
+            {
+                if (resource.Symbol.DeclaringSyntax is ResourceDeclarationSyntax declarationSyntax
+                    && declarationSyntax.TryGetBody()?.TryGetPropertyByName(AzResourceTypeProvider.ResourceNamePropertyName) is ObjectPropertySyntax objectPropertySyntax
+                    && objectPropertySyntax.Value is SyntaxBase resourceNameSyntax)
+                {
+                    resourcesAndNames.Add((resource, resourceNameSyntax));
+                }
+            }
+
+            if (resourcesAndNames.Any())
+            {
+                string formattedSearchName = resourceNameExpression.ToText();
+                string? evaluatedSearchNameLiteral = TryGetEvaluatedStringLiteral(model, resourceNameExpression)?.stringValue;
+
+                foreach (var (resource, resourceName) in resourcesAndNames)
+                {
+                    // First try a formatted expression match
+                    if (resourceName.ToText().EqualsOrdinally(formattedSearchName))
+                    {
+                        yield return resource;
+                    }
+
+                    // Then literal values (if they both evaluate to literal values)
+                    if (evaluatedSearchNameLiteral is not null
+                        && TryGetEvaluatedStringLiteral(model, resourceName) is (string resourceNameLIteral, _, _)
+                        && evaluatedSearchNameLiteral.EqualsOrdinally(resourceNameLIteral))
+                    {
+                        yield return resource;
+                    }
+                }
+            }
+        }
+
+        private static readonly Regex IsRegexRegex = new("[.$^([\\]]", RegexOptions.Compiled);
+
+        public static IEnumerable<FunctionCallSyntaxBase> FindFunctionCallsByName(SemanticModel model, SyntaxBase root, string @namespace, string functionNameOrRegex)
+        {
+            bool isFunctionNameARegex = IsRegexRegex.IsMatch(functionNameOrRegex);
+            Regex? regex = isFunctionNameARegex ? new Regex(functionNameOrRegex) : null;
+
+            return SyntaxAggregator.Aggregate(
+                source: root,
+                seed: new List<FunctionCallSyntaxBase>(),
+                function: (accumulated, syntax) =>
+                {
+                    if (syntax is FunctionCallSyntaxBase
```

Comment:
Is this check required? SemanticModelHelper.TryGetFunctionInNamespace(..) has a check to see if symbol bound to syntax node is a FunctionSymbol. Isn't that sufficient?

Reply:
And then casts.  Yep, thanks.